src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -114,16 +116,19 @@ private BoundExpression RewriteStringConcatenation(SyntaxNode syntax, BinaryOper
                         var second = leftFlattened[1];
                         var third = leftFlattened[2];
                         var fourth = leftFlattened[3];
-                        result = RewriteStringConcatenationFourExprs(syntax, first, second, third, fourth);
+                        Debug.Assert(previousLocals.Count <= 2);
+                        result = RewriteStringConcatenationFourExprs(syntax, first, second, third, fourth, previousLocals);
                     }
                     break;
 
                 default:
+                    Debug.Assert(previousLocals.Count == 0);
```

Comment:
>Debug.Assert(previousLocals.Count == 0); [](http://example.com/codeflow?start=20&length=40)

This assert fails for the following unit-test. Once all failing asserts are disabled, compiler crashes in stack optimizer.
```
        [Fact]
        public void Test4()
        {
            var source = """
                public class C 
                { 
                    static void Main()
                    {
                        var a = "a";
                        var b = "b";
                        var c = "c";
                        var d = "d";
                        var e = 'e';
                                                
                        System.Console.WriteLine((a + b + c) + (d + e)); 
                    }
                }
                """;

            var comp = CreateCompilation(source, options: TestOptions.ReleaseExe, targetFramework: TargetFramework.Net80);

            var verifier = CompileAndVerify(compilation: comp, expectedOutput: RuntimeUtilities.IsCoreClr8OrHigherRuntime ? "abcde" : null, verify: /*ExecutionConditionUtil.IsCoreClr ? default : */Verification.Skipped).VerifyDiagnostics();

            //verifier.VerifyIL("C.Main", "");
        }
```